File: 787.cpp
#include <queue>

struct Solution {
    using state = pair<int, pair<int, int>>;
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<vector<pair<int, int>>> edgeList(n);
        for (auto &edge : flights){
            edgeList[edge[0]].push_back({edge[1], edge[2]});
        }
        priority_queue<state, vector<state>, greater<state>> q;
        vector<int> visited(n);
        for (int i = 0; i < n; ++i){
            visited[i] = k + 1;
        }
        q.push({0, {src, 0}});
        while (!q.empty()){
            state st = q.top();
            int cost = st.first, node = st.second.first, stops = st.second.second;
            q.pop();
            if (node == dst){
                return cost;
            }
            if (stops >= visited[node] || stops == k + 1){
                continue;
            }
            visited[node] = stops;
            for (auto &pr : edgeList[node]){
                q.push({cost + pr.second, {pr.first, stops + 1}});
            }
        }
        return -1;
    }
};

File: sol/787.py
import heapq as hq
from collections import defaultdict

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        edges = defaultdict(dict)
        for s, e, p in flights:
            edges[s][e] = p
        h = [(0, src, 0)]
        visited = dict()
        while len(h) != 0:
            cost, state, stops = hq.heappop(h)
            if state == dst:
                return cost
            if stops >= visited.get(state, k + 1):
                continue
            visited[state] = stops
            if stops == k + 1:
                continue
            for n, p in edges[state].items():
                hq.heappush(h, (cost + p, n, stops + 1))
        return -1

